FFmpegInterop/FFmpegInteropMSS.cpp

```diff
@@ -18,1020 +18,432 @@
-					{
-						avAudioCodecCtx = nullptr;
-						hr = E_FAIL;
-					}
-					else
-					{
-						// Detect audio format and create audio stream descriptor accordingly
-						hr = CreateAudioStreamDescriptor(forceAudioDecode);
-						if (SUCCEEDED(hr))
-						{
-							hr = audioSampleProvider->AllocateResources();
-							if (SUCCEEDED(hr))
-							{
-								audioStreamSelected = true;
-								audioSampleProvider->EnableStream();
-								m_pReader->SetAudioStream(audioStreamIndex, audioSampleProvider);
-							}
-						}
-
-						if (SUCCEEDED(hr))
-						{
-							// Convert audio codec name for property
-							hr = ConvertCodecName(avAudioCodec->name, &audioCodecName);
-						}
-					}
-				}
+				AVDictionary* optionsRaw{ options.release() };
+				int result{ av_dict_set(&optionsRaw, keyA.c_str(), valueA.c_str(), 0) };
+				options.reset(exchange(optionsRaw, nullptr));
+				THROW_HR_IF_FFMPEG_FAILED(result);
 			}
 		}
-		else if (!forceAudioDecode)
+
+		// Open the format context for the stream
+		AVFormatContext* formatContextRaw{ m_formatContext.release() };
+		AVDictionary* optionsRaw{ options.release() };
+		int result{ avformat_open_input(&formatContextRaw, uri, nullptr, &optionsRaw) }; // The format context is freed on failure
+		options.reset(exchange(optionsRaw, nullptr));
+		THROW_HR_IF_FFMPEG_FAILED(result);
+		m_formatContext.reset(exchange(formatContextRaw, nullptr));
+
+		if (options != nullptr)
 		{
-			// FFMPEG doesn't have the codec but we can try to output the encoded type
-			audioStreamIndex = av_find_best_stream(avFormatCtx, AVMEDIA_TYPE_AUDIO, -1, -1, nullptr, 0);
-			
-			if (audioStreamIndex != AVERROR_STREAM_NOT_FOUND)
-			{
-				// Create audio stream descriptor from parameters
-				hr = CreateAudioStreamDescriptorFromParameters(avFormatCtx->streams[audioStreamIndex]->codecpar);
-				if (SUCCEEDED(hr))
-				{
-					hr = audioSampleProvider->AllocateResources();
-					if (SUCCEEDED(hr))
-					{
-						audioStreamSelected = true;
-						audioSampleProvider->EnableStream();
-						m_pReader->SetAudioStream(audioStreamIndex, audioSampleProvider);
-					}
-				}
-			}
+			// Options is not null if there was an issue with the provided FFmpeg options such as an invalid key or value.
+			WINRT_ASSERT(options == nullptr);
+			// TODO: Log options that weren't found
 		}
 	}
 
-	if (SUCCEEDED(hr))
+	void FFmpegInteropMSS::InitFFmpegContext(_In_opt_ const FFmpegInterop::FFmpegInteropMSSConfig& config)
 	{
-		// Find the video stream and its decoder
-		AVCodec* avVideoCodec = nullptr;
-		videoStreamIndex = av_find_best_stream(avFormatCtx, AVMEDIA_TYPE_VIDEO, -1, -1, &avVideoCodec, 0);
-		if (videoStreamIndex != AVERROR_STREAM_NOT_FOUND && avVideoCodec)
+		THROW_HR_IF_FFMPEG_FAILED(avformat_find_stream_info(m_formatContext.get(), nullptr));
+
+		bool hasVideo{ false };
+		bool hasAudio{ false };
+
+		for (unsigned int i{ 0 }; i < m_formatContext->nb_streams; i++)

```

Comment:
>for (unsigned int i{ 0 }; i < m_formatContext->nb_streams; i++) [](start = 2, length = 63)

nit: There's a potential regression here in terms of default stream selection. Previously, we were calling av_find_best_stream to find the indices of the default audio / video streams to select. av_find_best_stream has logic to select the best default stream based on bitrate / disposition, whereas we're now selecting the first audio + video stream that we enumerate. We could potentially retain the default stream selection behavior for audio / video streams by adding logic after this loop to perform selection based on the results from av_find_best_stream.